### PR TITLE
feat(bundle): add param and cred validation

### PR DESCRIPTION
**Pull request by vdice**
* adds Parameter and Credential validation, per spec
* validates both when validating a Bundle
* moved a few other validation functions closer to the structs they validate

### CODE REVIEW

File: bundle/bundle.go [hunk 1]
```diff
@@ -119,6 +119,16 @@ func (img *InvocationImage) DeepCopy() *InvocationImage {
 	return &img2
 }
 
+// Validate the image contents.
+func (img InvocationImage) Validate() error {
+	switch img.ImageType {
+	case "docker", "oci":
+		return validateDockerish(img.Image)
```
Comment (silvin-lubecki) [inline bundle/bundle.go]: Note that as we already vendor `docker/distribution`, we can use https://github.com/docker/distribution/blob/master/reference/reference.go#L189 instead to parse image reference?

File: bundle/bundle.go [hunk 2]
```diff
@@ -128,6 +138,15 @@ type Location struct {
 	EnvironmentVariable string `json:"env,omitempty" yaml:"env,omitempty"`
 }
 
+// Validate the Location
+func (l Location) Validate() error {
+	forbiddenPath := "/cnab/app/outputs"
+	if strings.HasPrefix(l.Path, forbiddenPath) {
+		return fmt.Errorf("Path must not be a subpath of %q", forbiddenPath)
```
Comment (silvin-lubecki) [inline bundle/bundle.go]: Maybe add the actual `path` to the error message? And I think error message shouldn't start with an upper case, as they can be wrapped.

File: bundle/bundle.go (hunk 1 above)
Comment (radu-matei) [reply bundle/bundle.go]: I'm not sure whether this should necessarily be part of this PR, but I agree we should have more robust validation.

Comment (silvin-lubecki) [reply bundle/bundle.go]: I can totally be in a follow up, as `validateDockerish` wasn't introduced in this PR.

File: bundle/bundle.go (hunk 2 above)
Comment (vdice) [reply bundle/bundle.go]: Good call; added.

File: bundle/bundle.go (hunk 1 above)
Comment (vdice) [reply bundle/bundle.go]: Good call; filed https://github.com/cnabio/cnab-go/issues/196

**Comment by vdice** [comment]
Updated/rebased; PTAL when convenient @silvin-lubecki .

**Pull request merged**